src/nvim/marktree.c

```diff
@@ -1821,40 +1822,54 @@ bool marktree_splice(MarkTree *b, int32_t start_line, int start_col, int old_ext
   return moved;
 }
 
-void marktree_move_region(MarkTree *b, int start_row, colnr_T start_col, int extent_row,
+void marktree_move_region(buf_T *buf, int start_row, colnr_T start_col, int extent_row,
                           colnr_T extent_col, int new_row, colnr_T new_col)
 {
+  MarkTree *b = buf->b_marktree;
   MTPos start = { start_row, start_col }, size = { extent_row, extent_col };
+  MTPos new = { new_row, new_col };
   MTPos end = size;
   unrelative(start, &end);
   MarkTreeIter itr[1] = { 0 };
   marktree_itr_get_ext(b, start, itr, false, true, NULL);
   kvec_t(MTKey) saved = KV_INITIAL_VALUE;
+  kvec_t(MTKey) signs = KV_INITIAL_VALUE;
   while (itr->x) {
     MTKey k = marktree_itr_current(itr);
     if (!pos_leq(k.pos, end) || (k.pos.row == end.row && k.pos.col == end.col
                                  && mt_right(k))) {
       break;
     }
+    if (k.flags & MT_FLAG_DECOR_SIGNTEXT) {
+      int row2 = marktree_get_altpos(b, k, NULL).row;
+      buf_signcols_invalidate_range(buf, k.pos.row, row2 < 0 ? k.pos.row : row2, -1);
+      kv_push(signs, k);
+    }
     relative(start, &k.pos);
+    unrelative(new, &k.pos);
     kv_push(saved, k);
     marktree_del_itr(b, itr, false);
   }
+  buf_signcols_validate(curwin, buf, false);
```

Comment:
Using `curwin` here looks kind of weird but I think its fine as this is only called with `curbuf` currently. So could also get rid of the `buf` argument I guess.

Reply:
This sounds like the typical which will hit us in the back of the head two years later when we least expect it. if the function depends on the window, it should take `win_T *` arg instead of buffer.

Alternatively, break out the part which is about other things than the actual marktree (to change.c or something like that), so this can keep only `MarkTree *` arg.

Reply:
> Alternatively, break out the part which is about other things than the actual marktree (to change.c or something like that), so this can keep only MarkTree * arg.

That would require to dupe the find/move marks in the range logic. Or for `marktree_move_region()` to return/fill a passed in `kvec_t(MTPair)`. Either of those preferred to adding `win_T *`?

Reply:
`win_T *` is fine for now. Then it at least is explicit what is going on (and we can clean it up later).